Door Improvements: Welding to fix doors, Door Sound Manager implementation, and various bugfixes

Bobtheapple opened the pull request:
### Purpose
Adds repairing doors using the welding tool.  Consolidates sound effect code from the various door modules into a new Door Sound Manager. Fixes a number of door related bugs.

### Notes:
I loaded into the game for the first time in a while and a door looked at me funny, this'll show it who's boss.

### Changelog:
CL: [New] Welders can now repair doors when in Help intent.
CL: [Improvement] Door sounds are now handled more robustly to make future improvements easier.
CL: [Improvement] Walking through opening or closing doors is now handled better.
CL: [Balance] Default delay to move through doors that are opening/closing increased to 50% of opening/closing time (typically 300ms), up from a static 100ms.
CL: [Fix] Fixed a bug where clicking on a door rather than bumping into the door bypassed the delay to move through altogether.
CL: [Fix] Fixed a bug where crowbars and jaws of life wouldn't always attack doors when using the harm intent.
CL: [Fix] Fixed a bug where crowbars could close doors even if they were bolted open.
CL: [Fix] Fixed a bug where powerful entities that can pry open doors couldn't close doors.
CL: [Fix] Fixed a bug where some interactions with doors would falsely say the the door was jammed.
CL: [Fix] Fixed a bug where welding doors would not use the proper object name in messages.
CL: [Fix] Fixed a bug where jaws of life could open welded doors, leaving the weld sprite floating in the air.
CL: [Fix] Fixed a bug where crowbars sometimes ignored welds on doors that were bolted.
CL: [Fix] Fixed a bug where using a welder with disarm intent on a door would do nothing, now opens the door.
CL: [Fix] Fixed a bug where doors could be interacted with while their animations were still playing.
CL: [Fix] Fixed a bug where floor tiles could be interacted with under a closed door by clicking rapidly.
CL: [Fix] Fixed a bug where door's lights would flash when pried opened by force regardless of power state.
CL: [Fix] Fixed a bug where crowbars and jaws of life could sometimes play the same sound as each other when prying doors.

There's probably a pile of other more obscure bugs that were fixed to, but these are the most obvious ones.

### Issues Addressed
#10196 - Resolved
#6980 & #6911 - Partially Resolved (the pry sound is implemented and the framework is there, but a placeholder sound is being used for the sound that plays when a door is finally pried open, I'll add the requested SFX in a separate PR that deals with addressables, wanted to limit this PR to strictly code changes)

Bobtheapple committed:
Added door repairing and fixed various door interaction bugs

Features
	Welders can now repair doors when in Help intent.

Bug Fixes
	Fixed a bug where crowbars and jaws of life wouldn't always attack doors when using the harm intent.
	Fixed a bug where crowbars could close doors even if they were bolted open.
	Fixed a bug where powerful entities that can pry open doors sometimes couldn't close doors.
	Fixed a bug where some interactions with doors would falsely say the the door was jammed.
	Fixed a bug where welding doors would not use the proper object name in messages.
	Fixed a bug where jaws of life could open welded doors.
	Fixed a bug where crowbars sometimes ignored welds on doors that were bolted.
	Fixed a bug where using a welder with disarm intent on a door would do nothing, now opens the door.
	Fixed a bug where doors could be interacted with while their animations were still playing.
	Fixed a bug where floor tiles could be interacted with under a closed door by clicking rapidly.

Bobtheapple committed:
Additional Door Bugfixes

Features
	Doors blocking players when opening/closing is now handled better.
	Additional tools for controlling door sounds have been added, new sounds will be implemented in a later PR.

Bug Fixes
	Fixed a bug where door's lights would flash when pryed opened by force.

Issues Resolved
	#10196

Bobtheapple committed:
Added Jaws of Life handling to the new DoorSoundController, added placeholder pry sound effect

Features
	Doors that are forced open now make a noise (noise to be adjusted in a later PR)

Bug Fixes
	Fixed a bug where Jaws of Life and Crowbar sometimes sounded alike when prying doors.

Addresses Issues
#6980
#6911

Bobtheapple committed:
Merge branch 'develop' of https://github.com/Bobtheapple/unitystation into develop

Bobtheapple committed:
Small Tweak to How opening door sounds are played

Bobtheapple committed:
Client Side Bugfix

Adjusted door animation code to work client side

Bobtheapple committed:
Tweak to the animation for slipping through the door early/late

Made it so the amount the door needs to be open can be set dynamically, rather than hard coded.

MaxIsJoe commented on UnityProject/Assets/Scripts/Objects/Doors/DoorSoundController.cs:
```diff
@@ -0,0 +1,134 @@
+            BoltsUp,
+		}
+
+        private AddressableAudioSource getSound(DoorSoundType sound)
```
All functions use CamelCase, so it should be `GetSound()`

MaxIsJoe commented on UnityProject/Assets/Scripts/Objects/Doors/Modules/WeldModule.cs:
```diff
@@ -27,51 +39,75 @@ public override void ClosedInteraction(HandApply interaction, HashSet<DoorProces
+					{
+						//Scale repair time based off of percent damage
+						//TODO use time multipliers from welder tools
+						float time = repairTime - repairTime * integrity.PercentageDamaged / 100f;
```
time is unused here?

MaxIsJoe reviewed (changes requested):
High quality PR, good work 👍 
just check these two things I pointed out.

Bobtheapple committed:
Bugfix

Properly implemented timescale factor for welding and cut its weight in half because welding was now too quick.  Fixed a function name to proper Camel Case.

MaxIsJoe reviewed (dismissed):
LGTM.
Merge after tests pass.

AtnerNT commented on UnityProject/Assets/Scripts/Objects/Doors/Modules/CrowbarModule.cs:
```diff
@@ -11,177 +11,162 @@ public class CrowbarModule : DoorModuleBase
+				if (interaction.HandObject == null
+					&& interaction.PerformerPlayerScript.PlayerTypeSettings.CanPryDoorsWithHands &&
+					(interaction.PerformerPlayerScript.TryGetComponent<AlienPlayer>(out var alienPlayer) == false ||
+					 alienPlayer.IsLarva == false))
```
Not something for this PR as aliens are there whole own bag of worms. But noting for future this shouldn't be hardcoded for Aliens/Alien Larvae. Would be better for aliens to just alter the CanPryDoorsWithHands directly on a stage by stage basis

Bobtheapple replied:
Yeah, this is how it was already implemented in the original CrowbarModule code with a "TODO: Fix this later".  I am not familiar enough with the current state of the Xenos code so I decided to let it be for now.

AtnerNT commented on UnityProject/Assets/Scripts/Objects/Doors/Modules/CrowbarModule.cs:
```diff
@@ -11,177 +11,162 @@ public class CrowbarModule : DoorModuleBase
+				if (interaction.HandObject == null
+					&& interaction.PerformerPlayerScript.PlayerTypeSettings.CanPryDoorsWithHands &&
+					(interaction.PerformerPlayerScript.TryGetComponent<AlienPlayer>(out var alienPlayer) == false ||
+					 alienPlayer.IsLarva == false))
```
Same thing here

AtnerNT commented on UnityProject/Assets/Scripts/Objects/Doors/Modules/CrowbarModule.cs:
```diff
@@ -11,177 +11,162 @@ public class CrowbarModule : DoorModuleBase
+			if (master.IsPerformingAction) return;
+
+			//Refuse if its a crowbar and the door has power
+			else if ((crowbarRequiresNoPower && master.HasPower) &&
```
At the end of each of these if's you return. So they dont actually need to be an else-if chain as the further code won't execute anyways. Codacy notes this as redundant jumps.

Bobtheapple replied:
Fixed!

Bobtheapple committed:
Fixed some redundant jumps

Removed unnecessary else statements in crowbar module.
